### PR TITLE
Balance Disorder: Add SerialFallbackProvider, a few multi-network refactorings

### LINKED ISSUE #1015: Not able to make WETH swap
### Discord Discussion Link

https://discord.com/channels/808358975287722045/888500174685614090/945404583604273162

### What browsers are you seeing the problem on?

Chrome

### What were you trying to do?

The user wanted to swap WETH for USDT

### What did not work?

WETH Balance is showing but the to be received USDT qty isn't calculated. Also not able to click 'Get final quote'

image

### Version

Community Edition v0.11.2

### Relevant log output

<details>
```shell
background.js:77588 internal: request payload: {"id":"30","target":"tally-provider-bridge","request":{"method":"eth_accounts","params":[]}}
background.js:77596     at chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:91216:22
    at chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:5597:22
    at Array.map (<anonymous>)
    at Emittery.emit (chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:5595:50)
background.js:77587  🪵 Tally log @ 2022-02-21T20:18:16.836Z
background.js:77588 internal: request payload: {"id":"31","target":"tally-provider-bridge","request":{"method":"eth_chainId","params":[]}}
background.js:77596     at chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:91216:22
    at chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:5597:22
    at Array.map (<anonymous>)
    at Emittery.emit (chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:5595:50)
background.js:77587  🪵 Tally log @ 2022-02-21T20:18:16.836Z
background.js:77588 internal response: {id: '31', result: '1'}
background.js:77596     at chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:91222:24
background.js:77587  🪵 Tally log @ 2022-02-21T20:18:16.852Z
background.js:77588 internal response: {id: '30', result: Array(1)}
background.js:77596     at chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:91222:24
    at async Promise.all (index 0)
    at async Emittery.emit (chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:5595:11)
background.js:77587  🪵 Tally log @ 2022-02-21T20:18:18.220Z
background.js:77588 internal: request payload: {"id":"32","target":"tally-provider-bridge","request":{"method":"eth_accounts","params":[]}}
background.js:77596     at chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:91216:22
    at chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:5597:22
    at Array.map (<anonymous>)
    at Emittery.emit (chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:5595:50)
background.js:77587  🪵 Tally log @ 2022-02-21T20:18:18.222Z
background.js:77588 internal response: {id: '32', result: Array(1)}
background.js:77596     at chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:91222:24
    at async Promise.all (index 0)
    at async Emittery.emit (chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:5595:11)
background.js:77587  🪵 Tally log @ 2022-02-21T20:18:18.223Z
background.js:77588 internal: request payload: {"id":"33","target":"tally-provider-bridge","request":{"method":"eth_accounts","params":[]}}
background.js:77596     at chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:91216:22
    at chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:5597:22
    at Array.map (<anonymous>)
    at Emittery.emit (chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:5595:50)
background.js:77587  🪵 Tally log @ 2022-02-21T20:18:18.224Z
background.js:77588 internal response: {id: '33', result: Array(1)}
background.js:77596     at chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:91222:24
    at async Promise.all (index 0)
    at async Emittery.emit (chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:5595:11)
background.js:77587  🪵 Tally log @ 2022-02-21T20:18:18.225Z
background.js:77588 internal: request payload: {"id":"34","target":"tally-provider-bridge","request":{"method":"eth_accounts","params":[]}}
background.js:77596     at chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:91216:22
    at chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:5597:22
    at Array.map (<anonymous>)
    at Emittery.emit (chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:5595:50)
background.js:77587  🪵 Tally log @ 2022-02-21T20:18:18.226Z
background.js:77588 internal response: {id: '34', result: Array(1)}
background.js:77596     at chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:91222:24
    at async Promise.all (index 0)
    at async Emittery.emit (chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:5595:11)
background.js:77587  🪵 Tally log @ 2022-02-21T20:18:18.230Z
background.js:77588 internal: request payload: {"id":"35","target":"tally-provider-bridge","request":{"method":"eth_chainId","params":[]}}
background.js:77596     at chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:91216:22
    at chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:5597:22
    at Array.map (<anonymous>)
    at Emittery.emit (chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:5595:50)
background.js:77587  🪵 Tally log @ 2022-02-21T20:18:18.230Z
background.js:77588 internal response: {id: '35', result: '1'}
background.js:77596     at chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:91222:24
background.js:77587  🪵 Tally log @ 2022-02-21T20:18:18.231Z
background.js:77588 internal: request payload: {"id":"36","target":"tally-provider-bridge","request":{"method":"eth_call","params":[{"from":"0xfbdd6ed4469b7d9a65fbf8b7f4160fbbd2ad7aa9","to":"0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2","data":"0xdd62ed3e000000000000000000000000fbdd6ed4469b7d9a65fbf8b7f4160fbbd2ad7aa9000000000000000000000000def1c0ded9bec7f1a1670819833240f027b25eff"},"latest"]}}
background.js:77596     at chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:91216:22
    at chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:5597:22
    at Array.map (<anonymous>)
    at Emittery.emit (chrome-extension://ngadognehgpdejcamckkhkokpndpdhgm/background.js:5595:50)
background.js:77022 WebSocket is already in CLOSING or CLOSED state.
(anonymous) @ background.js:77022
send @ background.js:76999
send @ background.js:87916
routeSafeRPCRequest @ background.js:91286
(anonymous) @ background.js:91220
(anonymous) @ background.js:5597
emit @ background.js:5595
await in emit (async)
postMessage @ background.js:91176
request @ background.js:91114
(anonymous) @ background.js:90966
send @ background.js:91026
(anonymous) @ background.js:76764
(anonymous) @ background.js:76303
json_rpc_provider_awaiter @ background.js:76285
perform @ background.js:76743
(anonymous) @ background.js:75839
fulfilled @ background.js:74644
Promise.then (async)
step @ background.js:74657
fulfilled @ background.js:74644
Promise.then (async)
step @ background.js:74657
(anonymous) @ background.js:74659
base_provider_awaiter @ background.js:74641
call @ background.js:75833
(anonymous) @ background.js:8979
fulfilled @ background.js:8930
Promise.then (async)
step @ background.js:8943
(anonymous) @ background.js:8945
__awaiter @ background.js:8927
call @ background.js:8976
(anonymous) @ background.js:89582
fulfilled @ background.js:89339
Promise.then (async)
step @ background.js:89352
(anonymous) @ background.js:89354
contracts_lib_esm_awaiter @ background.js:89336
(anonymous) @ background.js:89568
(anonymous) @ background.js:95293
await in (anonymous) (async)
(anonymous) @ background.js:94687
(anonymous) @ background.js:66866
step @ background.js:65746
(anonymous) @ background.js:65690
(anonymous) @ background.js:65808
__async @ background.js:65790
(anonymous) @ background.js:66843
(anonymous) @ background.js:66906
(anonymous) @ background.js:65641
(anonymous) @ background.js:63010
dispatchResponse @ background.js:63569
(anonymous) @ background.js:63127
background.js:77587  ❌ Tally error @ 2022-02-21T20:19:19.734Z
background.js:77588 Error getting block prices from BlockNative Error: timeout (requestBody=null, requestMethod="GET", timeout=120000, url="https://api.blocknative.com/gasprices/blockprices", code=TIMEOUT, version=web/5.5.1)
    at Logger.makeError (background.js:11662:25)
    at background.js:77843:41
genericLogger @ background.js:77588
error @ background.js:77614
getBlockPrices @ background.js:82977
background.js:77022 WebSocket is already in CLOSING or CLOSED state.
(anonymous) @ background.js:77022
send @ background.js:76999
(anonymous) @ background.js:76764
(anonymous) @ background.js:76303
json_rpc_provider_awaiter @ background.js:76285
perform @ background.js:76743
(anonymous) @ background.js:75617
fulfilled @ background.js:74644
Promise.then (async)
step @ background.js:74657
(anonymous) @ background.js:74659
base_provider_awaiter @ background.js:74641
getGasPrice @ background.js:75615
(anonymous) @ background.js:8887
(anonymous) @ background.js:8826
__awaiter @ background.js:8808
getFeeData @ background.js:8884
getBlockPrices @ background.js:82980
background.js:77022 WebSocket is already in CLOSING or CLOSED state.
(anonymous) @ background.js:77022
send @ background.js:76999
(anonymous) @ background.js:76764
(anonymous) @ background.js:76303
json_rpc_provider_awaiter @ background.js:76285
perform @ background.js:76743
(anonymous) @ background.js:75907
(anonymous) @ background.js:74659
base_provider_awaiter @ background.js:74641
base_provider_awaiter.oncePoll @ background.js:75906
check @ background.js:16098
(anonymous) @ background.js:16131
poll @ background.js:16075
(anonymous) @ background.js:75906
fulfilled @ background.js:74644
Promise.then (async)
step @ background.js:74657
fulfilled @ background.js:74644
Promise.then (async)
step @ background.js:74657
fulfilled @ background.js:74644
Promise.then (async)
step @ background.js:74657
(anonymous) @ background.js:74659
base_provider_awaiter @ background.js:74641
_getBlock @ background.js:75887
getBlock @ background.js:75949
getBlockPrices @ background.js:82980
background.js:77022 WebSocket is already in CLOSING or CLOSED state.
(anonymous) @ background.js:77022
send @ background.js:76999
(anonymous) @ background.js:76764
(anonymous) @ background.js:76303
json_rpc_provider_awaiter @ background.js:76285
perform @ background.js:76743
(anonymous) @ background.js:75907
(anonymous) @ background.js:74659
base_provider_awaiter @ background.js:74641
base_provider_awaiter.oncePoll @ background.js:75906
check @ background.js:16098
(anonymous) @ background.js:16131
poll @ background.js:16075
(anonymous) @ background.js:75906
fulfilled @ background.js:74644
Promise.then (async)
step @ background.js:74657
fulfilled @ background.js:74644
Promise.then (async)
step @ background.js:74657
fulfilled @ background.js:74644
Promise.then (async)
step @ background.js:74657
(anonymous) @ background.js:74659
base_provider_awaiter @ background.js:74641
_getBlock @ background.js:75887
getBlock @ background.js:75949
(anonymous) @ background.js:8886
(anonymous) @ background.js:8826
__awaiter @ background.js:8808
getFeeData @ background.js:8884
getBlockPrices @ background.js:82980
background.js:77664          GET https://api.blocknative.com/gasprices/blockprices 429
(anonymous) @ background.js:77664
(anonymous) @ background.js:77644
geturl_awaiter @ background.js:77626
getUrl @ background.js:77648
(anonymous) @ background.js:77869
(anonymous) @ background.js:77708
web_lib_esm_awaiter @ background.js:77690
(anonymous) @ background.js:77865
_fetchData @ background.js:77957
lib_esm_fetchJson @ background.js:77997
getBlockPrices @ background.js:82952
getBlockPrices @ background.js:82975
(anonymous) @ background.js:87911
pollBlockPrices @ background.js:87910
handler @ background.js:87716
handleAlarm @ background.js:73981
background.js:77664          GET https://api.blocknative.com/gasprices/blockprices 429
(anonymous) @ background.js:77664
(anonymous) @ background.js:77644
geturl_awaiter @ background.js:77626
getUrl @ background.js:77648
(anonymous) @ background.js:77869
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
(anonymous) @ background.js:77708
web_lib_esm_awaiter @ background.js:77690
(anonymous) @ background.js:77865
_fetchData @ background.js:77957
lib_esm_fetchJson @ background.js:77997
getBlockPrices @ background.js:82952
getBlockPrices @ background.js:82975
(anonymous) @ background.js:87911
pollBlockPrices @ background.js:87910
handler @ background.js:87716
handleAlarm @ background.js:73981
background.js:77664          GET https://api.blocknative.com/gasprices/blockprices 429
(anonymous) @ background.js:77664
(anonymous) @ background.js:77644
geturl_awaiter @ background.js:77626
getUrl @ background.js:77648
(anonymous) @ background.js:77869
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
(anonymous) @ background.js:77708
web_lib_esm_awaiter @ background.js:77690
(anonymous) @ background.js:77865
_fetchData @ background.js:77957
lib_esm_fetchJson @ background.js:77997
getBlockPrices @ background.js:82952
getBlockPrices @ background.js:82975
(anonymous) @ background.js:87911
pollBlockPrices @ background.js:87910
handler @ background.js:87716
handleAlarm @ background.js:73981
background.js:77664          GET https://api.blocknative.com/gasprices/blockprices 429
(anonymous) @ background.js:77664
(anonymous) @ background.js:77644
geturl_awaiter @ background.js:77626
getUrl @ background.js:77648
(anonymous) @ background.js:77869
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
(anonymous) @ background.js:77708
web_lib_esm_awaiter @ background.js:77690
(anonymous) @ background.js:77865
_fetchData @ background.js:77957
lib_esm_fetchJson @ background.js:77997
getBlockPrices @ background.js:82952
getBlockPrices @ background.js:82975
(anonymous) @ background.js:87911
pollBlockPrices @ background.js:87910
handler @ background.js:87716
handleAlarm @ background.js:73981
background.js:77664          GET https://api.blocknative.com/gasprices/blockprices 429
(anonymous) @ background.js:77664
(anonymous) @ background.js:77644
geturl_awaiter @ background.js:77626
getUrl @ background.js:77648
(anonymous) @ background.js:77869
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
(anonymous) @ background.js:77708
web_lib_esm_awaiter @ background.js:77690
(anonymous) @ background.js:77865
_fetchData @ background.js:77957
lib_esm_fetchJson @ background.js:77997
getBlockPrices @ background.js:82952
getBlockPrices @ background.js:82975
(anonymous) @ background.js:87911
pollBlockPrices @ background.js:87910
handler @ background.js:87716
handleAlarm @ background.js:73981
background.js:77664          GET https://api.blocknative.com/gasprices/blockprices 429
(anonymous) @ background.js:77664
(anonymous) @ background.js:77644
geturl_awaiter @ background.js:77626
getUrl @ background.js:77648
(anonymous) @ background.js:77869
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
(anonymous) @ background.js:77708
web_lib_esm_awaiter @ background.js:77690
(anonymous) @ background.js:77865
_fetchData @ background.js:77957
lib_esm_fetchJson @ background.js:77997
getBlockPrices @ background.js:82952
getBlockPrices @ background.js:82975
(anonymous) @ background.js:87911
pollBlockPrices @ background.js:87910
handler @ background.js:87716
handleAlarm @ background.js:73981
background.js:77664          GET https://api.blocknative.com/gasprices/blockprices 429
(anonymous) @ background.js:77664
(anonymous) @ background.js:77644
geturl_awaiter @ background.js:77626
getUrl @ background.js:77648
(anonymous) @ background.js:77869
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
Promise.then (async)
step @ background.js:77706
fulfilled @ background.js:77693
```
</details>

### PR DESCRIPTION
**Pull request by Shadowfiend**
Landing these together as they were somewhat entangled. The first 3 commits
remove network hardcoding from a few places, and move redux blocks to a
separate `networks` slice.

The last commit introduces a `SerialFallbackProvider` wrapping a list of Ethers
providers (currently just Alchemy websocket and standard providers). This
provider handles exponential backoff and retries that work for WebSockets. This
layers on top of existing exponential backoff/retries that are done in
non-WebSocket providers. Non-WebSocket providers also support explicit backoff
hints in Ethers, so in general the expectation is that these will rarely
trigger the `SerialFallbackProvider`'s retry logic.

WebSockets detect disconnection and automatically attempt to reconnect (with the
aforementioned backoff). Additionally, all subscriptions are tracked and
restored after a reconnect or a provider fallback. WebSocket subscriptions,
which require a long- lived connection, are not restored on no-WebSocket
providers.

Still missing here, but for a separate PR, is slowly trying to return to the
first provider in the list (required to be a WebSocket provider) even while
other fallbacks are being used, since only with a WebSocket provider can we
restore all subscriptions.

One outcome of this is that we no longer distinguish between polling and
WebSocket providers in `ChainService`; this is now all tucked behind the
facade of the `SerialFallbackProvider`.

### Testing

Use the wallet all around. dApps, swaps, balance updates, and transaction
subscriptions are all affected by this, since it hits the core provider
infrastructure.

In addition, check that cutting off the provider connection and then restoring
it keeps things working. The easiest way to do this is to kill your wifi
connection for 10-20s, then turn it back on. This should disconnect the initial
WebSocket connection and cause it to reconnect.

Closes #1015.

### CODE REVIEW

File: background/main.ts [hunk 1]
```diff
@@ -141,6 +141,54 @@ const REDUX_MIGRATIONS: { [version: number]: Migration } = {
...
+            [blockHeight: number]: unknown
+          }
+        }
+      }
+    }
+
+    const oldState = prevState as OldState
+
+    const networks: NetworkState = {
+      evm: {
```
Comment (mhluongo) [inline background/main.ts:167]: Good idea

File: background/main.ts [hunk 2]
```diff
@@ -549,7 +597,7 @@ export default class Main extends BaseService<never> {
 
   async getAccountEthBalanceUncached(address: string): Promise<bigint> {
     const amountBigNumber =
-      await this.chainService.pollingProviders.ethereum.getBalance(address)
+      await this.chainService.providers.ethereum.getBalance(address)
```
Comment (mhluongo) [inline background/main.ts:600]: Excited about only have one

File: background/services/chain/serial-fallback-provider.ts [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+ */
+function waitAnd<T, E extends Promise<T>>(
+  waitMs: number,
+  fn: () => E
+): Promise<T> {
+  return new Promise((resolve) => {
+    // TODO setTimeout rather than browser.alarms here could mean this would
+    // hang when transitioning to a transient background page? Can we do this
+    // with browser.alarms?
+    setTimeout(() => {
```
Comment (mhluongo) [inline background/services/chain/serial-fallback-provider.ts:31]: I'm also uncomfortable with this, but don't have a better immediate solution that isn't overly tied to our extension environment.

**Review by mhluongo — COMMENTED** [review]
Concept ACK, this could be a big win. Going to put it through its paces later today

File: background/services/chain/serial-fallback-provider.ts (hunk 3 above)
Comment (Shadowfiend) [reply background/services/chain/serial-fallback-provider.ts:31]: Yeah this one is extra weird. My understanding of stare restoration in manifest v3 isn't complete enough to be sure about how much we should care if setTimeouts are lost between executions.

**Comment by mhluongo** [comment]
Swaps and dApp connections seem to work, dApp tx requests, not so much. Here's what I got trying to swap KEEP to ETH on Uniswap

Screenshot from 2022-02-25 21-33-50

Screenshot from 2022-02-25 21-32-51

**Comment by Shadowfiend** [comment]
Interesting. This looks like an issue looking up gas prices. Will poke it when I get cycles.

**Comment by Shadowfiend** [comment]
@mhluongo I don't think this is an issue with this PR. I believe @wojciech-turek had run into this before as well. If this PR had broken something I would expect it to be the *ability* to get gas prices, but what's happening here is that we're massaging gas prices wrong, which happens in the UI/redux instead.

My guess is the issue is with how we compute maxFee/maxPriorityFee:

https://github.com/tallycash/extension/blob/8a1afbffb2c0f488e6fdb73b135312e84333f6c8/background/redux-slices/transaction-construction.ts#L180-L188

In particular, the former uses the base fee times a local multiplier ([`MAX_FEE_MULTIPLIER`](https://github.com/tallycash/extension/blob/main/background/constants/network-fees.ts#L14-L18)), while the latter pulls directly from `estimatedPrices`. This is important because, when Blocknative is unavailable, max priority fees are hardcoded in `estimatedPrices`, and max fees are offset by that hardcoded amount so they are never lower. However, once they're processed further in Redux, that invariant isn't guaranteed. My guess is the base fee while you were testing dipped low enough to drop below the fixed priority fee in the non-Blocknative estimate.

**Comment by mhluongo** [comment]
Excellent news, I wanted to merge this sooner rather than later. I opened a follow-up issue for the gas stuff [here](https://github.com/tallycash/extension/issues/1052)

**Review by mhluongo — APPROVED** [review]
This looks great, and the only QA issue was from a prior bug. Big unlock here :ok_hand:

**Comment by cairomassimo** [comment]
> Aisde: This is why I still think we need per-PR QA @cairomassimo. It's a great PR, but complicated for anyone but the author to debug. If we merged this before catching this bug, we'd be `git bisecting` to do basic release QA.

@mhluongo being a bit cynical here, `git bisecting` over PR merges would require less QA time overall than QA-ing every single one.

My take is that we should lower the bar for `main` from "releasable" to "code is good/stable enough to base new work on top of it". Then we could have a `stable` branch which is kept in "releasable" state (it could be just a pointer lagging behind `main`, that we update to the head of `main` once enough QA was done).

Keeping `main` releasable means that the QA-time grows super-linearly (quadratically?) with the size of the project: every PR potentially breaks a growing number of features (even if with low probability), that we should check before merge.
Instead, we could invest on preparing a test protocol, where we regularly and systematically test everything (that was touched) before each release. I think it would save some (dev) time in the end :)

That said, no problem in continuing this way!

**Pull request merged**